### DISCUSSION TITLE
openEMS on Ubuntu24.04

### DISCUSSION THREAD

**Discussion by istra711** [thliebig/openEMS-Project#268] [Q&A Usage]
Has anybody successfully installed openEMS 0.36 with python interface on Ubuntu24.04? I tried this several times but without any luck. The system python (3.12) requires now by default a virtual environment but that doesn't work somehow. I tried the "--break-system-packages" option but it didn't work because of dependencies problems with several python packages. To be able to use openEMS on Ubuntu 24.04 I have build a docker image based on Ubuntu 22.04 (python 3.10). This works so far with some modifications but my actual goal is to use openEMS with Blender and the new IntuitionRF plugin, unfortunately that seems to be impossible with a docker container. (Blender cannot see the openEMS installation inside of the container).

**Comment by LubomirJagos42** [comment]
hi, since I remember that openEMS python interface is written in cython and is for python 3.11 (also 3.10 could run as you wrote)
newest Blender is using python 3.12
if you want to use openEMS python interface you have to use older Blender version which is bundled with python 3.11, you will have to manualy copy that openEMS python modules folders from site-packages into Blender python site-packages

**Comment by thliebig** [comment]
@istra711 can you explain how you installed openEMS? Does it (except the python interface) build normally? 
I think the upate_openEMS.sh script does on use a python "venv" to build and install the python interface. 
Have you tried to setup a venv and then build the interface? It surely should work on any recent python version.

**Reply by istra711** [reply]
Hi Thorsten, thank you for your feedback.
Today I tried the installation again and did some documentation of my building steps.
BTW, I checked the "upate_openEMS.sh" but could not find any information regarding python and "venv".
For my installation and tests I am using a Linux-Mint22 VM from "osboxes.org". This VM is based on Ubuntu24.04.
Interesting to mention is the fact, that openEMS runs 15% faster on the Linux-VM (wmware) than on the windows host.
The installation was done as adviced by the openEMS wiki.
These are the steps I executed. (with some problems and solutions):

1) sudo apt-get install build-essential cmake git libhdf5-dev libvtk9-dev libboost-all-dev libcgal-dev libtinyxml-dev
2) sudo apt-get install qtbase5-dev libvtk9-qt-dev
3a) sudo apt-get install octave
3b) sudo apt-get install liboctave-dev ->> error because "liboctave-dev" doesn't exist in Ubuntu 24.04 repro!
Fix:
sudo apt-get install octave octave-dev

4) sudo apt-get install python3-numpy python3-matplotlib cython3 python3-h5py
5) git clone --recursive https://github.com/thliebig/openEMS-Project.git
6) cd openEMS-Project
7) sudo apt-get install gengetopt help2man groff pod2pdf bison flex libhpdf-dev libtool
8) ./update_openEMS.sh ~/opt/openEMS --with-hyp2mat --with-CTB --python

I received a "success message" but also the following error:
build openEMS and dependencies ... please wait
build successful, cleaning up tmp dir ...
build CSXCAD python module ... please wait
python module failed, please see logfile for more details...
build openEMS python module ... please wait
python module failed, please see logfile for more details...

In the log file I found the following message:
"ModuleNotFoundError: No module named 'setuptools'"

I was able to fix this by installing the a debian package:
sudo apt-get install python3-setuptools (and doing step 8 again)

9) Add to ~/.octaverc:
addpath('/home/osboxes/opt/openEMS/share/openEMS/matlab')
addpath('/home/osboxes/opt/openEMS/share/CSXCAD/matlab')
addpath('/home/osboxes/opt/openEMS/share/hyp2mat/matlab'); % hyp2mat package
addpath('/home/osboxes/opt/openEMS/share/CTB/matlab'); % circuit toolbox

First test with Octave and "Simple_Patch_Antenna.m" was successful but
"Bent_Patch_Antenna.m" stopped with errors:

setting up openEMS matlab/octave interface
compiling oct files
HDF5 library path found at: /usr/lib/x86_64-linux-gnu/hdf5/openmpi/libhdf5.so
/usr/lib/x86_64-linux-gnu/hdf5/serial
HDF5 include path found at: /usr/include/hdf5/serial
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/13/../../../x86_64-linux-gnu/Scrt1.o: in function `_start':
(.text+0x1b): undefined reference to `main'
collect2: error: ld returned 1 exit status
sh: 2: /usr/lib/x86_64-linux-gnu/hdf5/serial: Permission denied
error: running "setup" failed...
error: called from
    ReadHDF5Attribute at line 19 column 13
    ReadHDF5FieldData>ReadHDF5FieldData_octave at line 97 column 34
    ReadHDF5FieldData at line 29 column 19
    ReadHDF5Dump at line 31 column 7
    ConvertHDF5_VTK at line 51 column 15
    Bent_Patch_Antenna at line 167 column 1


This problem exists since quite some time and there is a fix mentioned here; 
https://github.com/thliebig/openEMS/pull/110
Is this an official fix now? 

A second possible solution was to copy the "h5readatt_octave.oct" from my old Ubuntu (were it was build without errors) to
the openEMS matlab folder, not sure if this is a good idea but it worked for me.

Now since I have a running setup of openEMS on Ubuntu 24.04 (Mint 22) I will try to install the lastest Blender and the IntuitionRF plug-in. BTW, it is possible to run Blender inside of a VM (very slow) but the 3D support of the VM must be disabled. 

Cheers,
Michael

**Reply by thliebig** [reply]
https://github.com/thliebig/openEMS/pull/110 was unfortunatly not the correct fix, it still works only in some cases
I guess I really have to take a closer look at fix this properly...
Additionally it seems for python you need to install  
`sudo apt install python3-setuptools`

**Reply by thliebig** [reply]
the "update_openEMS.sh" currently does not use a venv.
You could maybe run the tool as root? e.g.:
`sudo ./update_openEMS.sh /usr/local --with-hyp2mat --with-CTB --python`
This should install openEMS "globally" but I never tried it and it is not ideal...

**Reply by istra711** [reply]
Yes, I agree, the fix https://github.com/thliebig/openEMS/pull/110 is quite simple and probably will not work in every case. The "setup.m" has a problem in case it finds "libhdf5" in more than one location. In my case it comes with "serial" and "openmip". As a result "hdf5lib_dir" returns both paths and this generates an error. 
Using root might work but I don't like the global installation. With "sudo apt install python3-setuptools" and the modification of "setup.m" I am able to fix my problem. It other words, I think we can close the issue.

**Reply by thliebig** [reply]
What did you change about setup.m? Maybe you can come up with a good/save solution and either create a PR or just post it here and I can have a look at it and maybe commit it?

**Reply by av-p** [reply]
I encountered exactly the same problem. I've tried global installation -- this didn't work. [istra711](https://github.com/istra711) is right -- the problem is setup.m finds "libhdf5" in more than one location. Did a quick fix by overriding hdf5lib_dir and hdf5inc_dir in setup.m:
```
        if length(fn_so)>0 && length(fn_h)>0
            [hdf5lib_dir, hdf5lib_fn, ext] = fileparts(fn_so);
            hdf5lib_dir = '/usr/lib/x86_64-linux-gnu/hdf5/serial/'; % Override library path
            disp(["HDF5 library path found at: " hdf5lib_dir])
            [hdf5inc_dir, hdf5inc_fn, ext] = fileparts(fn_h);
            hdf5inc_dir = '/usr/include/hdf5/serial/'; % Override include path
            disp(["HDF5 include path found at: " hdf5inc_dir])
            mkoctfile("h5readatt_octave.cc", ["-L" hdf5lib_dir], ["-I" hdf5inc_dir], "-lhdf5")
        else

```
This helped in my case.

**Comment by nalyd88** [comment]
OpenEMS installed just fine for me today on 24.04. The key is simply to create and activate a virtual python environment first and then proceed with installation. I used uv to setup my virtual environment but you could use any thing you want. Once you have the environment active in your shell, install the required python packages, then re-run the openEMS installer script.

```bash
cd $HOME
uv venv --python 3.12
source .venv/bin/activate
uv pip install numpy h5py matplotlib cython setuptools

git clone --recursive https://github.com/thliebig/openEMS-Project.git
cd openEMS-Project
./update_openEMS.sh ~/.local --with-hyp2mat --with-CTB --python
```
